source/opt/interface_repair_pass.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "interface_repair_pass.h"
+#include "source/spirv_constant.h"
+namespace spvtools {
+namespace opt {
+InterfaceRepairPass::Status InterfaceRepairPass::Process() {
+  std::unordered_map<uint32_t, Function*> id2function_;
+  for (auto& fn : *get_module()) id2function_[fn.result_id()] = &fn;
+  bool modified = false;
+  for (auto& entry : get_module()->entry_points()) {
+    std::set<uint32_t> used_variables;
+    std::function<void(Function * func)> traverser = [&](Function* func) {
+      for (auto bi = func->begin(); bi != func->end(); ++bi) {
+        for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+          if (ii->opcode() == SpvOpFunctionCall) {
+            const auto called_id = ii->GetOperand(2).words[0];
+            traverser(id2function_.at(called_id));
+          } else {
+            for (uint32_t i = 0; i < ii->NumOperands(); ++i) {
+              auto& op = ii->GetOperand(i);
+              if (op.type == SPV_OPERAND_TYPE_ID) {
+                auto id = op.words[0];
+                if (used_variables.count(id)) continue;
+                auto* var = get_def_use_mgr()->GetDef(id);
+                if (!var || var->opcode() != SpvOpVariable) continue;
+                auto storage_class = var->GetSingleWordInOperand(0);
+                if (storage_class != SpvStorageClassFunction &&
+                    (get_module()->version() >= SPV_SPIRV_VERSION_WORD(1, 4) ||
```

Comment:
Hmm, I think varying variables may be removed even for older versions of SPIR-V.
For example, DXC lists much more varying variables than actually referenced. if you have 2 entry points in a HLSL file, each have an input variable, DXC will list both variables for both OpEntryPoint instructions.

Reply:
Getting the inputs and output wrong seems like a bug in DXC to me.  The interface should be determined by the HLSL source.  Can you give a sample HLSL file where this happens?  @jaebaek and I can take a look at it when we get time.  I'm still unsure about having a pass that modifies the inputs and outputs.

Reply:
This is my sample: microsoft/DirectXShaderCompiler#3759. 
The second comment.

Reply:
Sorry for this late reply.

The issue @thatname pointed is the case DXC lists inputs and outputs as operands of `OpEntryPoint`, but they are not actually used. First of all, we have to check whether we must remove the unused inputs and outputs from `OpEntryPoint` based on [SPIR-V spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpEntryPoint) or not.

> The set of Interface \<id> must be equal to or a superset of the global OpVariable Result \<id> referenced by the entry point’s static call tree, within the interface’s storage classes. Before version 1.4, the interface’s storage classes are limited to the Input and Output storage classes. Starting with version 1.4, the interface’s storage classes are all storage classes used in declaring all global variables referenced by the entry point’s call tree.

If we have to remove unused variables, I think it should be done in spirv-opt (in this PR), because it needs some analysis that is out of DXC's code generation.

I will first check https://github.com/microsoft/DirectXShaderCompiler/issues/3759 and get back to you.

Reply:
I've looked at the DXC issue.  That is a bug in DXC.  The issue is not about unused outputs.  The issue is that DXC add an output to the OpEntryPoint that has nothing to do with the entry point.  My guess is that DXC generates the instruction is a lazy way.  It adds all of the input and output variables in the module to the OpEntryPoint because it is assuming a single entry point per module.

I don't think that spirv-opt is the place to fix that.  DXC should generate the OpEntryPoint instruction based on the interface given in the entry point declaration.

Reply:
Sorry. I misunderstood it. Yeah we have to check the OpEntryPoint emission in DXC. I will get back after checking it.